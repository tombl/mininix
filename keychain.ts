import { assertEquals } from "@std/assert";
import { splitOnce } from "./util.ts";
import { decodeBase64 } from "@std/encoding/base64";

export const NIXOS_KEY =
  "cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=";

function parse(raw: string) {
  const [name, encodedBytes] = splitOnce(raw, ":");
  return { name, bytes: decodeBase64(encodedBytes) };
}

export type VerificationResult =
  | { valid: true }
  | { valid: false; reason: "INVALID_SIGNATURE" }
  | { valid: false; reason: "NO_SUCH_KEY" };

export class Keychain {
  #keys = new Map<string, { key: CryptoKey; raw: string }>();

  static async create(trustedKeys: string[]) {
    const keychain = new Keychain();
    await Promise.all(trustedKeys.map((raw) => keychain.trust(raw)));
    return keychain;
  }

  async trust(raw: string) {
    const { name, bytes } = parse(raw);

    const existing = this.#keys.get(name);
    if (existing) {
      assertEquals(
        existing.raw,
        raw,
        "Imported a key with the same name but different data",
      );
      return existing.key;
    }

    const key = await crypto.subtle.importKey(
      "raw",
      bytes,
      "Ed25519",
      true,
      ["verify"],
    );
    this.#keys.set(name, { key, raw });

    return key;
  }

  async verify(raw: string, data: BufferSource): Promise<VerificationResult> {
    const { name, bytes } = parse(raw);

    const signature = this.#keys.get(name);
    if (!signature) return { valid: false, reason: "NO_SUCH_KEY" };

    return (await crypto.subtle.verify("Ed25519", signature.key, bytes, data))
      ? { valid: true }
      : { valid: false, reason: "INVALID_SIGNATURE" };
  }
}
