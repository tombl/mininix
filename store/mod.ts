import type { NarListing } from "../nar.ts";
import type { NarInfo } from "../narinfo.ts";

export { BinaryCache } from "./binary-cache.ts";
export { FsCache } from "./fs-cache.ts";
export { MultiStore } from "./multi.ts";
export { createHttpHandler } from "./serve.ts";

export interface Store {
  storeDir: string;

  getInfo(
    hash: string,
    options?: { signal?: AbortSignal },
  ): Promise<NarInfo>;

  getListing(
    hash: string,
    options?: { signal?: AbortSignal },
  ): Promise<NarListing>;

  getNar(
    info: { narPathname: string },
    options?: { signal?: AbortSignal },
  ): Promise<ReadableStream<Uint8Array>>;
}

export interface WritableStore extends Store {
  putInfo(
    hash: string,
    info: NarInfo,
  ): Promise<void>;

  putListing(
    hash: string,
    listing: NarListing,
  ): Promise<void>;

  putNar(
    info: { narPathname: string },
    nar: ReadableStream<Uint8Array>,
  ): Promise<void>;
}

export function isWritableStore(store: Store): store is WritableStore {
  return "putInfo" in store;
}
