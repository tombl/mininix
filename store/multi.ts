import { assertEquals } from "@std/assert/equals";
import type { NarListing } from "../nar.ts";
import type { NarInfo } from "../narinfo.ts";
import { Data } from "../util.ts";
import type { Store } from "./mod.ts";

export class MultiStore extends Data<{ stores: Store[] }> implements Store {
  get storeDir(): string {
    const [{ storeDir }] = this.stores;
    for (const store of this.stores) {
      assertEquals(store.storeDir, storeDir);
    }
    return storeDir;
  }

  #indexes = new WeakMap<object, number>();
  getIndex(object: object): number | undefined {
    return this.#indexes.get(object);
  }

  hits: number[] = [];
  misses = 0;
  async #find<T extends object>(fn: (store: Store) => Promise<T>): Promise<T> {
    const errors = [];
    for (let i = 0; i < this.stores.length; i++) {
      const store = this.stores[i];
      try {
        const value = await fn(store);
        this.#indexes.set(value, i);
        this.hits[i] ??= 0;
        this.hits[i]++;
        return value;
      } catch (error) {
        errors.push(error);
      }
    }
    this.misses++;
    throw new AggregateError(errors, "all stores failed");
  }

  getInfo(
    hash: string,
    options?: { signal?: AbortSignal },
  ): Promise<NarInfo> {
    return this.#find((store) => store.getInfo(hash, options));
  }

  getListing(
    hash: string,
    options?: { signal?: AbortSignal },
  ): Promise<NarListing> {
    return this.#find((store) => store.getListing(hash, options));
  }

  getNar(
    info: { narPathname: string },
    options?: { signal?: AbortSignal },
  ): Promise<ReadableStream<Uint8Array>> {
    const index = this.getIndex(info);
    if (index === undefined) {
      return this.#find((store) => store.getNar(info, options));
    } else {
      return this.stores[index].getNar(info, options);
    }
  }
}
