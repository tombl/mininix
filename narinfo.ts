import { assert } from "@std/assert/assert";
import {
  CompressionAlgorithm,
  createDecompressionStream,
  isCompressionAlgorithm,
} from "./compression.ts";
import { Hash, LengthVerifierStream } from "./hash.ts";
import { Keychain } from "./keychain.ts";
import { createNarEntryStream } from "./nar.ts";
import { Store } from "./store.ts";
import { Data, parseKeyValue, ProgressReportingStream } from "./util.ts";

export class NarInfo extends Data<{
  store: Store;
  hash: string;

  storePath: string;
  narPathname: string;
  compression: CompressionAlgorithm;
  fileHash: Hash;
  narHash: Hash;
  fileSize: number;
  narSize: number;
  references: string[];
  deriver: string;
  sig: string;
}> {
  async files(
    options?: {
      signal?: AbortSignal;
      onProgress?: (current: number, total: number) => void;
    },
  ) {
    const [listing, nar] = await Promise.all([
      this.store.getListing(this.hash, options),
      this.store.getNar(this, options),
    ]);
    let body = nar.body!;

    if (options?.onProgress) {
      const { onProgress } = options;
      onProgress(0, this.fileSize);
      body = body.pipeThrough(
        new ProgressReportingStream((current) => {
          onProgress(current, this.fileSize);
        }),
      );
    }

    return body
      .pipeThrough(new LengthVerifierStream(this.fileSize))
      .pipeThrough(this.fileHash.createVerifierStream())
      .pipeThrough(createDecompressionStream(this.compression))
      .pipeThrough(new LengthVerifierStream(this.narSize))
      .pipeThrough(this.narHash.createVerifierStream())
      .pipeThrough(createNarEntryStream(listing));
  }

  verify(keychain: Keychain) {
    return keychain.verify(
      this.sig,
      new TextEncoder().encode(this.fingerprint()),
    );
  }

  fingerprint() {
    return [
      "1",
      this.storePath,
      this.narHash.raw,
      this.narSize.toFixed(0),
      this.references.map((ref) => `${this.store.storeDir}/${ref}`).join(","),
    ].join(";");
  }

  static parse(
    text: string,
    store: Store,
    hash: string,
  ) {
    const data = parseKeyValue(text, ": ");

    assert(
      isCompressionAlgorithm(data.Compression),
      `Unsupported compression algorithm: ${data.Compression}`,
    );

    return new NarInfo({
      store,
      hash,

      narPathname: data.URL,
      storePath: data.StorePath,
      compression: data.Compression,
      fileHash: new Hash(data.FileHash),
      narHash: new Hash(data.NarHash),
      fileSize: parseInt(data.FileSize),
      narSize: parseInt(data.NarSize),
      references: data.References.split(" ").filter(Boolean),
      deriver: data.Deriver,
      sig: data.Sig,
    });
  }

  get raw(): string {
    return [
      `URL: ${this.narPathname}`,
      `StorePath: ${this.storePath}`,
      `Compression: ${this.compression}`,
      `FileHash: ${this.fileHash.raw}`,
      `NarHash: ${this.narHash.raw}`,
      `FileSize: ${this.fileSize}`,
      `NarSize: ${this.narSize}`,
      `References: ${this.references.join(" ")}`,
      `Deriver: ${this.deriver}`,
      `Sig: ${this.sig}`,
    ].join("\n") + "\n";
  }
}
