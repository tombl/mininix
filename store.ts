import { assertEquals } from "@std/assert/equals";
import { isNarListing, NarListing } from "./nar.ts";
import { NarInfo } from "./narinfo.ts";
import { Data, parseKeyValue } from "./util.ts";

export interface Store {
  storeDir: string;

  getInfo(
    hash: string,
    options?: { signal?: AbortSignal },
  ): Promise<NarInfo>;

  getListing(
    hash: string,
    options?: { signal?: AbortSignal },
  ): Promise<NarListing>;

  getNar(
    info: { narPathname: string },
    options?: { signal?: AbortSignal },
  ): Promise<Response>;
}

export class BinaryCache extends Data<{
  url: URL;
  storeDir: string;
  wantMassQuery: boolean;
  priority: number;
}> implements Store {
  static async open(url: URL) {
    if (!url.pathname.endsWith("/")) url = new URL(url.href + "/");
    const response = await fetch(new URL("nix-cache-info", url));
    if (!response.ok) {
      throw new Error(
        `${response.status} ${response.statusText} at ${url}/nix-cache-info`,
      );
    }
    const data = parseKeyValue(await response.text(), ": ");
    return new BinaryCache({
      url,
      storeDir: data.StoreDir,
      wantMassQuery: data.WantMassQuery === "1",
      priority: parseInt(data.Priority ?? "0"),
    });
  }

  async #fetch(path: string, init?: RequestInit) {
    const url = new URL(path, this.url);
    const response = await fetch(url, init);
    if (!response.ok) {
      throw new Error(`${response.status} ${response.statusText} at ${url}`);
    }
    return response;
  }

  async getInfo(hash: string, options?: { signal?: AbortSignal }) {
    const response = await this.#fetch(hash + ".narinfo", options);
    return NarInfo.parse(await response.text(), this, hash);
  }

  async getListing(
    hash: string,
    options?: { signal?: AbortSignal },
  ): Promise<NarListing> {
    const response = await this.#fetch(hash + ".ls", options);
    const json: unknown = await response.json();
    if (!isNarListing(json)) throw new Error("invalid nar listing");
    return json;
  }

  getNar(
    info: { narPathname: string },
    options: { signal?: AbortSignal } = {},
  ) {
    return this.#fetch(info.narPathname, options);
  }
}

export class MultiStore extends Data<{ stores: Store[] }> implements Store {
  get storeDir() {
    const [{ storeDir }] = this.stores;
    for (const store of this.stores) {
      assertEquals(store.storeDir, storeDir);
    }
    return storeDir;
  }

  #indexes = new WeakMap<object, number>();
  getIndex(object: object) {
    return this.#indexes.get(object);
  }

  hits: number[] = [];
  misses = 0;
  async #find<T extends object>(fn: (store: Store) => Promise<T>): Promise<T> {
    const errors = [];
    for (let i = 0; i < this.stores.length; i++) {
      const store = this.stores[i];
      try {
        const value = await fn(store);
        this.#indexes.set(value, i);
        this.hits[i] ??= 0;
        this.hits[i]++;
        return value;
      } catch (error) {
        errors.push(error);
      }
    }
    this.misses++;
    throw new AggregateError(errors, "all stores failed");
  }

  getInfo(
    hash: string,
    options?: { signal?: AbortSignal },
  ): Promise<NarInfo> {
    return this.#find((store) => store.getInfo(hash, options));
  }

  getListing(
    hash: string,
    options?: { signal?: AbortSignal },
  ): Promise<NarListing> {
    return this.#find((store) => store.getListing(hash, options));
  }

  getNar(
    info: { narPathname: string },
    options?: { signal?: AbortSignal },
  ): Promise<Response> {
    const index = this.getIndex(info);
    if (index === undefined) {
      return this.#find((store) => store.getNar(info, options));
    } else {
      return this.stores[index].getNar(info, options);
    }
  }
}
