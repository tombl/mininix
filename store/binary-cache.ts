import { isNarListing, type NarListing } from "../nar.ts";
import { NarInfo } from "../narinfo.ts";
import { Data, parseKeyValue } from "../util.ts";
import type { Store } from "./mod.ts";

export class BinaryCache extends Data<{
  url: URL;
  storeDir: string;
  wantMassQuery: boolean;
  priority: number;
}> implements Store {
  static async open(url: URL): Promise<BinaryCache> {
    if (!url.pathname.endsWith("/")) url = new URL(url.href + "/");
    const response = await fetch(new URL("nix-cache-info", url));
    if (!response.ok) {
      throw new Error(
        `${response.status} ${response.statusText} at ${url}/nix-cache-info`,
      );
    }
    const data = parseKeyValue(await response.text(), ": ");
    return new BinaryCache({
      url,
      storeDir: data.StoreDir,
      wantMassQuery: data.WantMassQuery === "1",
      priority: parseInt(data.Priority ?? "0"),
    });
  }

  async #fetch(path: string, init?: RequestInit) {
    const url = new URL(path, this.url);
    const response = await fetch(url, init);
    if (!response.ok) {
      throw new Error(`${response.status} ${response.statusText} at ${url}`);
    }
    return response;
  }

  async getInfo(
    hash: string,
    options?: { signal?: AbortSignal },
  ): Promise<NarInfo> {
    const response = await this.#fetch(hash + ".narinfo", options);
    return NarInfo.parse(await response.text(), this, hash);
  }

  async getListing(
    hash: string,
    options?: { signal?: AbortSignal },
  ): Promise<NarListing> {
    const response = await this.#fetch(hash + ".ls", options);
    const json: unknown = await response.json();
    if (!isNarListing(json)) throw new Error("invalid nar listing");
    return json;
  }

  async getNar(
    info: { narPathname: string },
    options: { signal?: AbortSignal } = {},
  ): Promise<ReadableStream<Uint8Array>> {
    const response = await this.#fetch(info.narPathname, options);
    return response.body!;
  }
}
