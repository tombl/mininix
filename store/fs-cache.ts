import { assert } from "@std/assert/assert";
import { join } from "@std/path/join";
import { isNarListing, type NarListing } from "../nar.ts";
import { NarInfo } from "../narinfo.ts";
import type { WritableStore } from "./mod.ts";

export class FsCache implements WritableStore {
  storeDir = "/nix/store";

  #dir: string;
  private constructor(dir: string) {
    this.#dir = dir;
  }

  static async open(dir: string): Promise<FsCache> {
    await Deno.mkdir(join(dir, "nar"), { recursive: true });
    return new FsCache(dir);
  }

  async putInfo(hash: string, info: NarInfo) {
    const path = join(this.#dir, hash + ".narinfo");
    await Deno.writeTextFile(path, info.raw);
  }
  async getInfo(
    hash: string,
    options?: { signal?: AbortSignal },
  ): Promise<NarInfo> {
    const path = join(this.#dir, hash + ".narinfo");
    const text = await Deno.readTextFile(path, options);

    return NarInfo.parse(text, this, hash);
  }

  async putListing(hash: string, listing: NarListing) {
    const path = join(this.#dir, hash + ".ls");
    await Deno.writeTextFile(path, JSON.stringify(listing));
  }
  async getListing(
    hash: string,
    options?: { signal?: AbortSignal },
  ): Promise<NarListing> {
    const path = join(this.#dir, hash + ".ls");
    const json: unknown = JSON.parse(await Deno.readTextFile(path, options));
    assert(isNarListing(json));
    return json;
  }

  async putNar(
    info: { narPathname: string },
    nar: ReadableStream<Uint8Array>,
  ) {
    assert(info.narPathname.startsWith("nar/"));
    const path = join(this.#dir, info.narPathname);
    try {
      await Deno.writeFile(path + ".tmp", nar);
    } catch (error) {
      await Deno.remove(path + ".tmp");
      throw error;
    }
    await Deno.rename(path + ".tmp", path);
  }
  async getNar(
    info: { narPathname: string },
    options?: { signal?: AbortSignal },
  ): Promise<ReadableStream<Uint8Array>> {
    assert(info.narPathname.startsWith("nar/"));
    const path = join(this.#dir, info.narPathname);
    const body = await Deno.open(path, { read: true });

    options?.signal?.addEventListener("abort", () => body.close());

    // TODO: check that closing the body closes the file
    return body.readable;
  }
}
