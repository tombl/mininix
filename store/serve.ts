import { isNarListing } from "../nar.ts";
import { NarInfo } from "../narinfo.ts";
import { FsCache } from "./fs-cache.ts";
import { isWritableStore, type Store } from "./mod.ts";

export function createHttpHandler(
  store: Store,
  priority: number,
): (request: Request) => Promise<Response | null> {
  return async (request) => {
    const { pathname } = new URL(request.url, "http://host.invalid");

    if (!["GET", "HEAD", "PUT"].includes(request.method)) {
      return new Response("method not allowed", { status: 405 });
    }

    if (pathname === "/nix-cache-info") {
      if (request.method === "PUT") {
        return new Response("method not allowed", { status: 405 });
      }

      return new Response(
        [
          `StoreDir: ${store.storeDir}`,
          "WantMassQuery: 1",
          `Priority: ${priority}`,
        ].join("\n"),
        { headers: { "content-type": "text/x-nix-cache-info" } },
      );
    }

    if (pathname.endsWith(".narinfo")) {
      const hash = pathname.slice(1, -".narinfo".length);

      if (request.method === "PUT") {
        if (!isWritableStore(store)) {
          return new Response("store is not writable", { status: 405 });
        }

        const raw = await request.text();
        let info;
        try {
          info = NarInfo.parse(raw, store, hash);
        } catch (error) {
          return new Response(
            `invalid narinfo: ${error}`,
            { status: 400 },
          );
        }

        return await store.putInfo(hash, info).then(
          () => new Response("ok", { status: 200 }),
          (error: unknown) => {
            console.error(error);
            return new Response(`failed to store: ${error}`, { status: 500 });
          },
        );
      }

      return await store.getInfo(hash).then(
        (info) =>
          new Response(info.raw, {
            headers: { "content-type": "text/x-nix-narinfo" },
          }),
        () => new Response("nar not found", { status: 404 }),
      );
    }

    if (pathname.endsWith(".ls")) {
      const hash = pathname.slice(1, -".ls".length);

      if (request.method === "PUT") {
        if (!isWritableStore(store)) {
          return new Response("store is not writable", { status: 405 });
        }

        const listing = await request.json();
        if (!isNarListing(listing)) {
          return new Response("invalid listing", { status: 400 });
        }

        return await store.putListing(hash, listing).then(
          () => new Response("ok", { status: 200 }),
          (error: unknown) => {
            console.error(error);
            return new Response(`failed to store: ${error}`, { status: 500 });
          },
        );
      }

      return await store.getListing(hash).then(
        (listing) =>
          new Response(JSON.stringify(listing), {
            headers: { "content-type": "application/json" },
          }),
        () => new Response("listing not found", { status: 404 }),
      );
    }

    if (pathname.startsWith("/nar/")) {
      const info = { narPathname: pathname.slice(1) };

      if (request.method === "PUT") {
        if (!isWritableStore(store)) {
          return new Response("store is not writable", { status: 405 });
        }

        return await store.putNar(info, request.body!).then(
          () => new Response("ok", { status: 200 }),
          (error: unknown) => {
            console.error(error);
            return new Response(`failed to store: ${error}`, { status: 500 });
          },
        );
      }

      return await store.getNar(info).then(
        (nar) =>
          new Response(nar, {
            headers: { "content-type": "application/x-nix-nar" },
          }),
        () => new Response("nar not found", { status: 404 }),
      );
    }

    return null;
  };
}

if (import.meta.main) {
  const store = await FsCache.open("cache");
  const handler = createHttpHandler(store, 1);
  Deno.serve(async (request) =>
    (await handler(request)) ?? new Response("not found", { status: 404 })
  );
}
