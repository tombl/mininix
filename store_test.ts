import { assertEquals, assertRejects } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { Keychain, NIXOS_KEY } from "./keychain.ts";
import { NarInfo } from "./narinfo.ts";
import { BinaryCache } from "./store.ts";

const netPermission = await Deno.permissions.query({ name: "net" });

Deno.test({
  name: "nixpkgs#hello",
  ignore: netPermission.state !== "granted",
  async fn(t) {
    const cache = await BinaryCache.open(new URL("https://cache.nixos.org"));
    const nixosKeychain = await Keychain.create([NIXOS_KEY]);
    const emptyKeychain = new Keychain();

    let info: NarInfo;

    await t.step("fetch", async (t) => {
      info = await cache.narInfo("kwmqk7ygvhypxadsdaai27gl6qfxv7za");
      await assertSnapshot(t, info);
    });

    await t.step("verify", async () => {
      assertEquals(await info.verify(nixosKeychain), { valid: true });
      assertEquals(await info.verify(emptyKeychain), {
        valid: false,
        reason: "NO_SUCH_KEY",
      });
    });

    await t.step("nar", async (t) => {
      await t.step("valid", async (t) => {
        const validInfo = info.clone();
        const nar = await validInfo.nar();
        const bytes = await new Response(nar).arrayBuffer();
        await assertSnapshot(t, bytes.byteLength);
      });

      await t.step("wrong compressed size", async () => {
        const wrongInfo = info.clone();
        wrongInfo.fileSize *= 2;
        await assertRejects(async () => {
          const nar = await wrongInfo.nar();
          await new Response(nar).arrayBuffer();
        });
      });

      await t.step("wrong decompressed size", async () => {
        const wrongInfo = info.clone();
        wrongInfo.narSize *= 2;
        await assertRejects(async () => {
          const nar = await wrongInfo.nar();
          await new Response(nar).arrayBuffer();
        });
      });

      await t.step("wrong compressed hash", async () => {
        const wrongInfo = info.clone();
        wrongInfo.fileHash.hash[0]++;
        await assertRejects(async () => {
          const nar = await wrongInfo.nar();
          await new Response(nar).arrayBuffer();
        });
      });

      await t.step("wrong decompressed hash", async () => {
        const wrongInfo = info.clone();
        wrongInfo.narHash.hash[0]++;
        await assertRejects(async () => {
          const nar = await wrongInfo.nar();
          await new Response(nar).arrayBuffer();
        });
      });
    });
  },
});
