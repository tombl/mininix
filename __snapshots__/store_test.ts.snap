export const snapshot = {};

snapshot[`nixpkgs#hello > fetch 1`] = `
NarInfo {
  compression: "xz",
  deriver: "pznj731mjim1xdd5mir97l20pk3gy5a8-hello-2.12.1.drv",
  fileHash: Hash {
    algorithm: "sha256",
    hash: Uint8Array(32) [
      154,
      223,
      186,
      27,
      240,
      59,
      27,
      55,
      0,
      67,
      192,
      157,
      136,
      86,
      236,
      93,
      152,
      47,
      31,
      174,
      13,
      35,
      42,
      193,
      30,
      102,
      242,
      182,
      224,
      95,
      6,
      141,
    ],
    raw: "sha256:1386bzhbdwk63v0jl8qdmqgjz62xxib8i7f08c03f6rvy0dvmpws",
  },
  fileSize: 52484,
  hash: "a7hnr9dcmx3qkkn8a20g7md1wya5zc9l",
  narHash: Hash {
    algorithm: "sha256",
    hash: Uint8Array(32) [
      21,
      39,
      148,
      161,
      76,
      100,
      230,
      236,
      237,
      255,
      33,
      25,
      222,
      213,
      254,
      229,
      223,
      1,
      191,
      49,
      177,
      96,
      143,
      22,
      41,
      32,
      31,
      201,
      69,
      133,
      60,
      131,
    ],
    raw: "sha256:10rwhm2wj7r054b8yq5i66zh3pz5zvaxw691zznyrrk49jhr89qm",
  },
  narPathname: "nar/1386bzhbdwk63v0jl8qdmqgjz62xxib8i7f08c03f6rvy0dvmpws.nar.xz",
  narSize: 234680,
  references: [
    "65h17wjrrlsj2rj540igylrx7fqcd6vq-glibc-2.40-36",
    "a7hnr9dcmx3qkkn8a20g7md1wya5zc9l-hello-2.12.1",
  ],
  sig: "cache.nixos.org-1:HBx8vRSFoiXhIKtXLv7LSHdEcIH5Ji4zeADslQEKWn+j++VJpWM8ASIMQ4ilCucWkzN1SlZdqSLpNhHwWPayDA==",
  store: BinaryCache {
    priority: 40,
    storeDir: "/nix/store",
    url: URL {
      hash: "",
      host: "cache.nixos.org",
      hostname: "cache.nixos.org",
      href: "https://cache.nixos.org/",
      origin: "https://cache.nixos.org",
      password: "",
      pathname: "/",
      port: "",
      protocol: "https:",
      search: "",
      username: "",
    },
    wantMassQuery: true,
  },
  storePath: "/nix/store/a7hnr9dcmx3qkkn8a20g7md1wya5zc9l-hello-2.12.1",
}
`;
